Convert rewriting and proof-mode chapters to prodn (#13470)
coq/coq: doc/sphinx/proofs/writing-proofs/rewriting.rst

=========================
Reasoning with equalities
=========================

There are multiple notions of :gdef:`equality`:

- :gdef:`Liebniz equality` is the standard
  way to define equality in Coq and the Calculus of Inductive Constructions, which is in

>   terms of a relation.  The standard library defines `eq` this way:

Author: Maybe you should tell me.  Is `eq` considered a mathematical relation, a proposition or what?  `eq` is a `Prop`.  I was thinking of a relation R(a, b) such that if R contains (a, b) then a and b are equal.  Probably not phrased correctly.

`eq` is generally bound to the notation `a = b`, right?

>   terms of a relation.  The standard library defines `eq` this way:

Reviewer: Right. And `eq` is a binary relation and applied to its arguments it is a proposition. "contains" is probably not appropriate since it refers to the way relations are usually defined in set theory (as sets, more precisely as subsets of the cartesian product of two sets for the case of a binary relation), whereas in type theory relations are functions (whose output is a proposition).

>   terms of a relation.  The standard library defines `eq` this way:

Reviewer: Anyway, this intro now looks very good to me!

>   terms of a relation.  The standard library defines `eq` this way:

Author: I changed to use the right terminology:

```which is in terms of a binary relation, i.e. a binary function that returns `True` if its arguments are equal and otherwise `False`.```

>   terms of a relation.  The standard library defines `eq` this way:

Author: Is the True/False part right?

>   terms of a relation.  The standard library defines `eq` this way:

Reviewer: No. This formulation is not correct. This is a function that "returns" a proposition (in the sense of the type after the last arrow), but not a function that computes (thus it doesn't actually ever "returns" any value such as `True` / `False`---which are not even primitive Coq values BTW).

Said otherwise, while `eq` is a binary relation, `eq x y` (i.e. `eq` fully applied) is a proposition. A proposition can be proved to be true, proved to be false, impossible to prove true or false, etc.

>    Inductive eq (A:Type) (x:A) : A -> Prop := eq_refl : x = x

Author: Yup.  I couldn't remember its name.  I think this is the first I've used it.

>    .. todo PR: the definition in .v file is a little different

Author: Check what I wrote.

- :gdef:`Setoid equality <setoid equality>` defines equality in terms of an equivalence
  relation.  A :gdef:`setoid` is a set that is equipped with an equivalence relation
  (see https://en.wikipedia.org/wiki/Setoid).  These are needed to form a :gdef:`quotient set`
  or :gdef:`quotient`
  (see https://en.wikipedia.org/wiki/Equivalence_Class).  In Coq, users generally work
  with setoids rather than constructing quotients, for which there is no specific support

>   (see also `#10871 <https://github.com/coq/coq/issues/10871>`_).

Reviewer: This issue has probably too much discussion and bad karma to be advertised in the refman I think.

Author: I wondered about that.  While the PR is public, there's no need to give it more visibility.  Took out the "See ...".

> - :gdef:`Definitional equality <definitional equality>` is equality based on the conversion rules, which

Reviewer: Probably worth linking to the "Conversion rules" section in the "Core language" chapter.

  Coq can determine automatically.  When two terms are definitionally equal, Coq knows it can
  replace one with the other, such as with :tacn:`change` `X with Y`, among many
  other advantages.

>   .. todo PR: A and B are convertible iff they are definitionally equal??

Reviewer: Yes. And some error messages use the word "not convertible" instead of "not definitionally equal" BTW.

.. _rewritingexpressions:

Rewriting expressions
---------------------

These tactics use the equality :g:`eq : forall {A : Type}, A -> A -> Prop` defined in
file ``Logic.v`` (see :ref:`coq-library-logic`). The notation for :g:`eq t u` is

> :g:`t = u`.

Reviewer: This is becoming less useful since it repeats part of the introduction. We could remove it and instead change the title of the section to "Rewriting with Leibniz or setoid equality" and make the second section "Rewriting with definitional equality" start at `change`.

.. tacn:: rewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }

   .. insertprodn oriented_rewriter one_term_with_bindings

   .. prodn::
      oriented_rewriter ::= {? {| -> | <- } } {? @natural } {? {| ? | ! } } @one_term_with_bindings
      one_term_with_bindings ::= {? > } @one_term {? with @bindings }

.. todo PR: This description is wrong.  This doesn't work:
   Goal forall x: nat, x = 2 -> x = 0 /\ x = 1.
   intro. intro H.
   rewrite (forall (x:nat), eq x 3).
   But rewrite H. does.

>    Also want to check if finds subterms inside forall, thought I had an example where it didn't.

Reviewer: You are correct. It is wrong to say that `one_term` must have the following form. It is its type that must have this form. Thus, typically this `one_term` will be a hypothesis or theorem name.

The answer to your second question is that `rewrite` cannot find subterms inside forall, but `setoid_rewrite` can.

Author: I suppose everything shown in the context is a type, but it seems to me referring to `type of` here isn't very clear because we don't generally use that language.  We just refer to a hypothesis.  Does that make sense?

```
1 subgoal
x : nat
H : x = 2
______________________________________(1/1)
x = 0 /\ x = 1
```

Reviewer: Everything after a colon is a type. `H` is the name of a hypothesis but also a (very simple) term. Its type is `x = 2`. We tend to confuse the hypothesis term / name (`H`) and type (`x = 2`) indeed!

Author: I'd like to add a couple paragraphs somewhere that show the current goal and explain the terms "hypothesis" "conclusion" "current goal" "shelved goal" etc. and mention that the part after the : and below the line is actually a type.  Where do you suggest?  In Proof Handling?

Reviewer: Yes, absolutely! Note that I've recently renamed this file into `proof-mode.rst` and I was hoping to change the title as well. What you propose would be a very nice introduction to this chapter.

Reviewer 2: > The answer to your second question is that rewrite cannot find subterms inside forall, but setoid_rewrite can.

This isn't quite accurate; rewrite will find all and only the subterms of a forall that don't mention the forall binders.  So with `x = y`, you can still rewrite `forall a, x = a`.

Author: Thanks.  I have this now:

```[rewrite] won't find occurrences inside `forall` that refer to variables bound by the `forall` ```

   Note that :n:`eq @term__1 @term__2` is typically written with the infix notation
   :n:`@term__1 = @term__2`.  You must `Require Setoid` to use the tactic
   with a setoid equality or with :ref:`setoid rewriting <generalizedrewriting>`.
   In the general form, any :n:`@binder` may be used, not just :n:`(x__i: A__i)`.

   .. todo doublecheck the @binder comment is correct.

   :n:`rewrite @one_term` finds subterms matching :n:`@term__1` in the goal,
   and replaces them with :n:`@term__2` (or the reverse if `<-` is given).
   Some of the variables :g:`x`\ :sub:`i` are solved by unification,
   and some of the types :n:`A__1, ..., A__n` may become new
   subgoals.  It won't find occurrences inside `forall` that refer
   to variables bound by the `forall`; use :tacn:`setoid_rewrite`
   if you want to find such occurrences.

   :n:`{+, @oriented_rewriter }`
     The :n:`@oriented_rewriter`\s are applied sequentially
     to the first goal generated by the previous :n:`@oriented_rewriter`.  If any of them fail,
     the tactic fails.

   :n:`{? {| -> | <- } }`
     For `->` (the default), :n:`@term__1` is rewritten

>      into :n:`@term__2`.  For `<-`, :n:`@term__2` is rewritten into :n:@`term__1`.

Reviewer: ```suggestion
     into :n:`@term__2`.  For `<-`, :n:`@term__2` is rewritten into :n:`@term__1`.
```

     is the maximum number of rewrites to perform; otherwise :n:`natural` is the exact number
     of rewrites to perform.

>      '?' (without :n:`@natural) performs the rewrite as many times as possible

Reviewer: This is making the build fail:

```suggestion
     '?' (without :n:`@natural`) performs the rewrite as many times as possible
```

     (possibly zero times).

>      This form never fails.  '!' (without :n:`@natural) performs the rewrite as many

Reviewer: Again the same issue:

```suggestion
     This form never fails.  '!' (without :n:`@natura`l) performs the rewrite as many
```

Reviewer: Still failing because there was a typo in my second fix.

  which is in terms of a binary relation, i.e. a binary function that returns
  a `Prop`.  The standard library
  defines `eq` similar to this:

   .. coqdoc::

      Inductive eq {A : Type} (x : A) : A -> Prop := eq_refl : eq x x.

  The notation `x = y` represents the term `eq x y`.  The notation `x = y :> A`
  gives the type of x and y explicitly.

- :gdef:`Setoid equality <setoid equality>` defines equality in terms of an equivalence
  relation.  A :gdef:`setoid` is a set that is equipped with an equivalence relation
  (see https://en.wikipedia.org/wiki/Setoid).  These are needed to form a :gdef:`quotient set`
  or :gdef:`quotient`
  (see https://en.wikipedia.org/wiki/Equivalence_Class).  In Coq, users generally work
  with setoids rather than constructing quotients, for which there is no specific support.

- :gdef:`Definitional equality <definitional equality>` is equality based on the
  :ref:`conversion rules <Conversion-rules>`, which Coq can determine automatically.
  When two terms are definitionally equal, Coq knows it can
  replace one with the other, such as with :tacn:`change` `X with Y`, among many

>   other advantages.  ":term:`Convertible <convertible>`" is another way of saying that

Reviewer: We could also add a reference from the place where "convertible" is defined to here. I would simply change the `or *equivalent*` phrase over there to ``` or :term:`definitionally equal <definitional equality>` ```.

  two terms are definitionally equal.

.. _rewritingexpressions:

Rewriting with Leibniz and setoid equality
------------------------------------------

.. tacn:: rewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }

   .. insertprodn oriented_rewriter one_term_with_bindings

   .. prodn::
      oriented_rewriter ::= {? {| -> | <- } } {? @natural } {? {| ? | ! } } @one_term_with_bindings
      one_term_with_bindings ::= {? > } @one_term {? with @bindings }

>    Rewrites terms based on equalities.  :n:`@one_term` is a term whose type has the form:

Reviewer: "one_term is a term" seems a bit obvious. What about "The type of one_term should have the form"?

      :n:`{? forall {+ (x__i: A__i) } , } EQ @term__1 @term__2`

   where :g:`EQ` is the Leibniz equality `eq` or a registered setoid equality.
   Note that :n:`eq @term__1 @term__2` is typically written with the infix notation
   :n:`@term__1 = @term__2`.  You must `Require Setoid` to use the tactic

>    with a setoid equality or with setoid rewriting.  In the general form,

Reviewer: ```suggestion
   with a setoid equality or with :ref:`setoid rewriting <generalizedrewriting>`.  In the general form,
```

   any :n:`@binder` may be used, not just :n:`(x__i: A__i)`.

   .. todo doublecheck the @binder comment is correct.

   :n:`rewrite @one_term` finds subterms matching :n:`@term__1` in the goal,
   and replaces them with :n:`@term__2` (or the reverse if `<-` is given).
   Some of the variables :g:`x`\ :sub:`i` are solved by unification,
   and some of the types :n:`A__1, ..., A__n` may become new
   subgoals.  It won't find occurrences inside `forall` that refer
   to variables bound by the `forall`; use :tacn:`setoid_rewrite`
   if you want to find such occurrences.

   :n:`{+, @oriented_rewriter }`
     The :n:`@oriented_rewriter`\s are applied sequentially
     to the first goal generated by the previous :n:`@oriented_rewriter`.  If any of them fail,
     the tactic fails.

   :n:`{? {| -> | <- } }`

>      For `->` (the default), :n:`term__1` is rewritten
>      into :n:`term__2`.  For `<-`, :n:`term__2` is rewritten into :n:`term__1`.

Reviewer: ```suggestion
     For `->` (the default), :n:`@term__1` is rewritten
     into :n:`@term__2`.  For `<-`, :n:`@term__2` is rewritten into :n:`@term__1`.
```

   :n:`{? @natural } {? {| ? | ! } }`

>      :n:`natural` is the number of rewrites to perform.  If `?` is given, :n:`@natural`
>      is the maximum number of rewrites to perform; otherwise :n:`natural` is the exact number

Reviewer: ```suggestion
     :n:`@natural` is the number of rewrites to perform.  If `?` is given, :n:`@natural`
     is the maximum number of rewrites to perform; otherwise :n:`@natural` is the exact number
```

>      of rewrites to perform.

Reviewer: Maybe specify that `natural!` means the same as `natural`?

>      '?' (without :n:`@natural`) performs the rewrite as many times as possible
>      (possibly zero times).
>      This form never fails.  '!' (without :n:`@natural`) performs the rewrite as many

Reviewer: For consistency with the previous paragraph (could also be changed in the other way in the previous paragraph):

```suggestion
     `?` (without :n:`@natural`) performs the rewrite as many times as possible
     (possibly zero times).
     This form never fails.  `!` (without :n:`@natural`) performs the rewrite as many
```

     times as possible
     and at least once.  The tactic fails if the requested number of rewrites can't
     be performed.

   :n:`@occurrences`
     If :n:`@occurrences` specifies multiple occurrences, the tactic succeeds if
     any of them can be rewritten.  If not specified, only the first occurrence
     in the conclusion is replaced.

>      If :n:`at @occs_nums` is specified, rewriting is always done with setoid
>      rewriting, even for Leibniz’s equality.

Reviewer: ```suggestion
     If :n:`at @occs_nums` is specified, rewriting is always done with
     :ref:`setoid rewriting <generalizedrewriting>`, even for Leibniz’s equality.
```

     .. todo not finding occurrences within forall is not a general limitation of

>         occurrences.  Probably there are other tactics that have this limitation.

Reviewer: Is this todo really worth keeping in the code?

   :n:`by @ltac_expr3`
     If specified, is used to resolve all side conditions generated by the tactic.

   .. exn:: Tactic failure: Setoid library not loaded.
      :undocumented:

      .. todo You can use Typeclasses Debug to tell whether rewrite used
         setoid rewriting.  Example here: https://github.com/coq/coq/pull/13470#discussion_r539230973

   .. exn:: The @term provided does not end with an equation.

>       :undocumented:

Reviewer: After grepping "does not end" and "with an equation", I conclude that this error message doesn't exist anymore. The error I got while trying to apply something of the wrong type was "Cannot find a relation to rewrite."

   .. exn:: Tactic generated a subgoal identical to the original goal.
      :undocumented:

   .. exn:: Found no subterm matching @term in @ident.

>             Found no subterm matching @term the current goal.

Reviewer: ```suggestion
            Found no subterm matching @term in the current goal.
```

=========================
Reasoning with equalities
=========================

There are multiple notions of :gdef:`equality` in Coq:

- :gdef:`Leibniz equality` is the standard
  way to define equality in Coq and the Calculus of Inductive Constructions, which is in
  terms of a relation.  The standard library defines `eq` similar to this:

   .. coqdoc::

      Inductive eq {A : Type} (x : A) : A -> Prop := eq_refl : eq x x.

   The notation `x = y` is equivalent to `eq x y`.  The notation `x = y :> A`
   gives the type of x and y explicitly.

- :gdef:`Setoid equality <setoid equality>` defines equality in terms of an equivalence
  relation.  A :gdef:`setoid` is a set that is equipped with an equivalence relation
  (see https://en.wikipedia.org/wiki/Setoid).  These are needed to form a :gdef:`quotient set`
  or :gdef:`quotient`
  (see https://en.wikipedia.org/wiki/Equivalence_Class).  In Coq, users generally work
  with setoids rather than constructing quotients, for which there is no specific support.

> - :gdef:`Definitional equality <definitional equality>` is equality based on the conversion
>   rules, which Coq can determine automatically.  (See :ref:`Conversion-rules`)

Reviewer: The link doesn't need to be shown prominently. I think this reads better like this:

```suggestion
- :gdef:`Definitional equality <definitional equality>` is equality based on the
  :ref:`conversion rules <Conversion-rules>`, which Coq can determine automatically.
```

  When two terms are definitionally equal, Coq knows it can
  replace one with the other, such as with :tacn:`change` `X with Y`, among many
  other advantages.  ":term:`Convertible <convertible>`" is another way of saying that
  two terms are definitionally equal.

>   .. todo PR: A and B are convertible iff they are definitionally equal??

Reviewer: You forgot to remove the `todo PR`.

.. _rewritingexpressions:

Rewriting with Leibniz and setoid equality
------------------------------------------

.. tacn:: rewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }

   .. insertprodn oriented_rewriter one_term_with_bindings

   .. prodn::
      oriented_rewriter ::= {? {| -> | <- } } {? @natural } {? {| ? | ! } } @one_term_with_bindings
      one_term_with_bindings ::= {? > } @one_term {? with @bindings }

   .. todo PR: When I search for "rewrite (" in the .v files, I see tons of examples that I don't

>       understand.  Need to say something about them.

Reviewer: Yes, the current text is not explaining that properly. You can give the name of a lemma or hypothesis (this is the most common case) but you can also give any term whose type is of the right form (shown below). The typical use case will be to apply a lemma to its argument explicitly (also the `with` clause also supports this).

Author: Getting there.  Explain me these:

```
rewrite (not_not_iff P) in H by dec
rewrite (add_lt_mono_r _ _ n)          (is there an implicit parm?)
rewrite (CR_of_Q_plus R 1 1)           (also implicit?)
```

Maybe we should mention `_` as a wildcard, too?

Author: If you provide a term, do you have the prove the equality in a side condition (side goal)?

Reviewer: > Getting there. Explain me these:
> 
> ```
> rewrite (not_not_iff P) in H by dec
> rewrite (add_lt_mono_r _ _ n)          (is there an implicit parm?)
> rewrite (CR_of_Q_plus R 1 1)           (also implicit?)
> ```

It's not clear what you are asking, but basically, we pass some parameters to these lemmas (which Coq may or may not have inferred by itself). The `_` wildcard allows skipping some parameters which we know Coq will correctly infer.

> Maybe we should mention `_` as a wildcard, too?

It's not specific to `rewrite` though.

> If you provide a term, do you have the prove the equality in a side condition (side goal)?

No. It works exactly the same as if you provide a hypothesis or a lemma (which are terms too). The type of the term must have the expected form (i.e. `forall ..., EQ term1 term2`). The proof of the equality is the term itself (a proof term).

Author: The examples are not in the form we described, `(EQ term1 term2)`.  The first has only one parameter, the others have three.  What do these examples mean and how are they interpreted?  Is there some kind of evaluation applied before doing the rewrite?

Reviewer: > The examples are not in the form we described, `(EQ term1 term2)`. The first has only one parameter, the others have three.

It's not the provided term but its type which must have the form `forall ..., EQ term1 term2`. This is the case in these examples. This is exactly the same as when giving lemma or hypothesis names (it's the type of `H` which must be in the form `forall ..., EQ term1 term2`).

Author: That needs some explanation.  It's not at all obvious and it's new to me.  Dropping in "type of a term" is just hiding the details.  How many tactics rely on the type of the term?

Reviewer: > How many tactics rely on the type of the term?

The answer is basically all. This is very fundamental stuff so very important that you understand fully. We can move the discussion somewhere else (or have a call) if needed.

I'll try to explain once more using `rewrite` and `apply` as examples:

If we forget the (non-important here) considerations on opaque vs transparent definitions, a lemma and a definition are the same thing (that's why they are presented in the same section https://coq.inria.fr/refman/language/core/definitions.html): they associate a name to a term (the body) and its type. If the type is a proposition, then we call the body a "proof term".

When we use a lemma in a proof with `rewrite lemma_name` or `apply lemma_name`, we usually don't care what the proof term (body) of this lemma is. What we care about is its type (the statement of the lemma). In particular, in the case of `apply` the conclusion of the statement (the type thus) should be unified with the current goal, while in the case of `rewrite`, the conclusion of the statement should be an equality (or an equivalence) between two terms, and the left-hand side will be unified with subterms of the current goal.

Both tactics do not actually take `qualid` as argument but terms (`one_term` to be specific on the parsing level). This means that instead of using `lemma_name`, you can also use its proof term directly. But whether you use `lemma_name` or its proof term, what you actually care about is not the term but its type (the statement). This doesn't depend on whether you use a lemma or a term directly.

To take a concrete example, if `foo` is a lemma whose statement (type) is `forall x y z : nat, x = y -> y = z -> x = z`, then `foo a b` is a (proof) term whose type is `forall z : nat, a <= b -> b <= z -> a <= z`, thus you may use to replace a goal of the form `a <= z` with two subgoals `a <= b` and `b <= z`.

Author: Most of that makes sense.  A hypothesis has a type but not a value (nor a proof term), right? A bit different from the usual programming language notions.

The last paragraph bears a little more explanation.  The last paragraph is an example of partial application, right?  Is this the only way (or the most common way) to do partial application for a proof term?  Is that the dual form of partial application to a function, e.g.

Fixpoint add2 m := add 2 m.
Compute (add2 5).

We should explain partial application somewhere in the doc but probably not in this file.

Author: This didn't work, nor does `(plus_Sn_m 3)`.  Where am I going astray?

```
Print plus_Sn_m.
Goal True.
let t := (plus_Sn_m 3 4) in idtac t.

(* The reference plus_Sn_m was not found in the current environment. *)
```

Reviewer: > Most of that makes sense. A hypothesis has a type but not a value (nor a proof term), right?

Right.

>  The last paragraph is an example of partial application, right?

Right.

>  Is this the only way (or the most common way) to do partial application for a proof term? Is that the dual form of partial application to a function

Yes, because these proof terms *are* functions, as `Print` shows.

> This didn't work, nor does `(plus_Sn_m 3)`. Where am I going astray?

Here is your fix: `let t := constr:(plus_Sn_m 3 4) in idtac t.`
The problem was just that the `let` construct of Ltac does not expect a term by default. So it was not looking in the right place of the environment...

Author: Thanks.  For `let t := constr:(plus_Sn_m) in idtac t.`, I thought I'd see the proof term.  Am I not seeing it because the proof is opaque?   Is there a way to see it so I can explain it in the text?

Reviewer: Yes, for instance `let t := eval unfold plus_Sn_m in plus_Sn_m in idtac t.` fails with `plus_Sn_m is opaque.`

I think only `Print` lets you see through opaque proofs.

Author: I'd still like to explain how `rewrite (CR_of_Q_plus R 1 1)` works for this tactic--still don't get it.   It's defined in a Strucure as:

`    CR_of_Q_plus : forall q r : Q, CReq (CR_of_Q (q+r))
                                           (CRplus (CR_of_Q q) (CR_of_Q r));
`

Reviewer: Because it's defined in a `Structure` (i.e. a record), it has additional arguments that do not appear on this line (the `R`). Thus `CR_of_Q_plus R 1 1` has the type `CReq (CR_of_Q (1+1)) (CRplus (CR_of_Q 1) (CR_of_Q 1))`. `CReq` is a setoid equality and thus you can rewrite with it.

Author: Yeah, that makes sense.  I was thinking that `one_term` had to be an already-proven statement, which is probably wrong--when will `rewrite` introduce introduce `one_term` as a side condition?  And I take it that being in a structure doesn't mean that `CR_of_Q_plus` is proven?

(For "replace", I see the language "This equality is automatically solved if it occurs among the hypotheses, or if its symmetric form occurs.")

Reviewer: Even if you did not declare it as a lemma, the simple fact that you can pass around a term means that its type is inhabited and thus proven. Stuff coming from a structure can thus be "proven".

`rewrite` will only generate side-conditions if the lemma that is used has preconditions that are not automatically resolved.

Example:

```coq
Goal 1 = 1.
rewrite <- (min_l 1 1).
```

generates two goals `Nat.min 1 1 = Nat.min 1 1` and the side-condition `1 <= 1`.

For `replace ... with ...`, the side-condition will be the equality itself if it cannot be found in the hypotheses. The confusing bit might be that both `replace` and `rewrite` take a `one_term` as argument but it hasn't the same status at all. In the case of `replace`, it is the subterm to be replaced, while in the case of `rewrite`, it is a proof of an equality or equivalence.

Author: > Even if you did not declare it as a lemma, the simple fact that you can pass around a term means that its type is inhabited and thus proven

Ah, OK.  Still need to look at more tactics and consider whether they are using terms or the type of term.  IIRC you said most (all?) tactics are using the type of the term.

How does this work if you pass unprovable terms such as `1 = 0` or `True -> False`?

Reviewer 2: > How does this work if you pass unprovable terms such as 1 = 0 or True -> False?

Their type is `Prop`, which is inhabited.  (Of course it doesn't make sense to rewrite with a lemma which proves that `Prop` is inhabited, so using these particular "proof terms" really only makes sense with something like `pose proof`.)

Reviewer 2: >  IIRC you said most (all?) tactics are using the type of the term.

Most tactics use the term itself to build the proof term, which means that only the type is seen in the goal / context.  The exceptions include `change` (which leaves no evidence in the proof term), `replace` and Ltac `match` (which use the terms you pass them for matching), `pose` (which uses the same term in the proof term evidence and in the body of the variable added to the context), and `assert`/`cut`, which do not take a proof and instead ask you to prove a thing of the type you give them. As usual, I think `refine` is a useful primitive to think about here, since pretty much every other tactic bottoms out in a call to `refine`.

   Rewrites terms based on equalities.  :n:`@one_term` can be the name of a hypothesis that has

>    the form :n:`EQ @term__1 @term__2` or a lemma that has the form:

Reviewer: I would avoid making a distinction between a hypothesis and a lemma. A lemma can be of the form `EQ term1 term2` if the terms are grounded (no free variable). This doesn't happen a lot but can exist. And a hypothesis can have binders (`forall ...`) just as a lemma.

Author: I think I like it this way:

![image](https://user-images.githubusercontent.com/1253341/101727600-9f171800-3a69-11eb-80e0-343084ab45be.png)

>    :n:`forall (x__1: A__1) ... (x__n: A__n), EQ @term__1 @term__2`

Reviewer: What about:
```suggestion
   :n:`{* forall {+ @binder }, } EQ @term__1 @term__2`
```

It would be more general, although it would need some rewrite in the following text.

Another option requiring less rewrite is:
```suggestion
   :n:`{* forall {+ (x__i: A__i) }, } EQ @term__1 @term__2`
```

Author: Are all the options of `binder` supported here?  I expect the common case is just `(x__i : A__i)`.  I have this now:

![image](https://user-images.githubusercontent.com/1253341/101728100-a1c63d00-3a6a-11eb-9fb9-3e93f49840a9.png)

I could just say that this is the most common form and that you can, in fact, use any type of `binder` in place of `(x__i : A__i)`.

Reviewer: > I could just say that this is the most common form and that you can, in fact, use any type of `binder` in place of `(x__i : A__i)`.

Sounds good to me (I think that's the truth).

   where :g:`EQ` is the Leibniz equality `eq` or a registered setoid equality.  (Typically
   the infix notation :n:`@term__1 = @term__2` is defined as :n:`eq @term__1 @term__2`.

>    Either form may used.)

Reviewer: I wouldn't write the parenthesis that way. I would rather say that `eq term1 term2` is usually written `term1 = term2` using the corresponding notation.

Author: Ah, I call that a "parenthetical comment".

   .. todo PR: I can't remember where the following sentence came from.  It seems

>       incorrect.  Maybe worth having some comparison to setoid_rewrite, though

Reviewer: Indeed, it is incorrect. `rewrite` can use the default implementation for Leibniz equality or the setoid rewrite implementation automatically. However, the user may also choose to call the latter explicitly even in the case of a Leibniz equality by using the `setoid_rewrite` tactic. If `rewrite` picks the setoid implementation, it will fail with an error message if the `Setoid` library hasn't been required first though.

A good way of finding out what implementation is used is to set the `Typeclasses Debug` flag (since setoid rewrite relies on the typeclass mechanism internally):

```coq
Require Setoid.

Set Typeclasses Debug.
Axiom foo : True <-> False.
Goal True.
rewrite foo.
(* Use the setoid implementation since the output is:
1: looking for (RelationClasses.subrelation iff (Basics.flip Basics.impl)) without backtracking
1.1: exact Morphisms.iff_flip_impl_subrelation on (RelationClasses.subrelation iff (Basics.flip Basics.impl)), 0 subgoal(s)
Finished run 1 of resolution.
*)
Abort.

Axiom bar : 0 = 1.
Goal 0 = 0.
rewrite bar.
(* Use the default implementation since there is no output. *)
Abort.

   and replaces them with :n:`@term__2` (or the reverse if `<-` is given).
   Some of the variables :g:`x`\ :sub:`i` are solved by unification,
   and some of the types :n:`A__1, ..., A__n` may become new
   subgoals.  It won't find occurrences inside `forall`; use :tacn:`setoid_rewrite`
   if you want to find such occurrences.

   :n:`{+, @oriented_rewriter }`
     The :n:`@oriented_rewriter`\s are applied sequentially
     to the first goal generated by the previous :n:`@oriented_rewriter`.  If any of them fail,
     the tactic fails.

   :n:`{? {| -> | <- } }`
     For `->` (the default), :n:`term__1` is rewritten
     into :n:`term__2`.  For `<-`, :n:`term__2` is rewritten into :n:`term__1`.

   :n:`{? @natural } {? {| ? | ! } }`
     :n:`natural` is the number of rewrites to perform.  If `?` is given, :n:`@natural`
     is the maximum number of rewrites to perform; otherwise :n:`natural` is the exact number
     of rewrites to perform.

     '?' performs the rewrite as many times as possible (possibly zero times).
     This form never fails.  '!' performs the rewrite as many times as possible
     and at least once.  The tactic fails if the requested number of rewrites can't

>      be performed.

Reviewer: Maybe clarify that this is when these symbols are used alone (without `natural`).

   :n:`@occurrences`
     If :n:`@occurrences` specifies multiple occurrences, the tactic succeeds if
     any of them can be rewritten.  If not specified, only the first occurrence
     in the conclusion is replaced.

     If :n:`at @occs_nums` is specified, rewriting is always done with setoid
     rewriting, even for Leibniz’s equality, so you must
     ``Require Setoid`` to use this form.

     .. todo PR occurrences are not found within forall.  Is this specific to

>         rewrite or a general limitation of occurrence finding?

Reviewer: It is probably not entirely specific to this tactic but not general either. It depends on the tactic.

Author: OK.  Is this a necessary condition or would it be worthwhile to relax the condition?

Reviewer: In rewrite, it's not strictly necessary since `setoid_rewrite` proves that it's possible to do differently (though it may be more complicated and may lead to error messages that are harder to understand). But I think that there certainly are tactics for which it just doesn't make sense to select an occurrence under binders.

          rewrite * {? {| -> | <- } } @one_term at @rewrite_occs in @ident {? by @ltac_expr3 }
   :name: rewrite (star); _
   :undocumented:

.. tacn:: rewrite_db @ident {? in @ident }
   :undocumented:

.. tacn:: replace @one_term__from with @one_term__to {? @occurrences } {? by @ltac_expr3 }
          replace {? {| -> | <- } } @one_term {? @occurrences }
   :name: replace; _

   The first form replaces all free occurrences of :n:`@one_term__from`
   in the current goal with :n:`@one_term__to` and generates an equality
   :n:`@one_term__to = @one_term__from`
   as a subgoal. (Note the generated equality is reversed with respect
   to the order of the two terms in the tactic syntax; see
   issue `#13480 <https://github.com/coq/coq/issues/13480>`_.)
   This equality is automatically solved if it occurs among
   the assumptions, or if its symmetric form occurs.

   The second form, with `->` or no arrow, replaces :n:`@one_term`
   with :n:`@one_term'` using
   the first assumption whose type has the form :n:`@term = @term’`.
   If `<-` is given, the tactic uses the first assumption with the reverse form,

>    i.e. :n:`@term’ = @term`.

Author: Where exactly do you get `term__to` from this form?

![image](https://user-images.githubusercontent.com/1253341/101730639-43e82400-3a6f-11eb-9e6e-fe8fe8d2d5c5.png)

>    i.e. :n:`@term’ = @term`.

Reviewer: >Perhaps it's that `one_term` is itself an equality?

No. As explained in the text, you find the first assumption whose type is `term__from = term__to` (where `term__from` is the one you give as an argument to `replace` and `term__to` is just whatever is on the right-hand side of the equality). Said otherwise, what the tactic is looking for is rather `term__from = ?x` where whatever is unified with `?x` will replace `term__from` in the selected occurrences.

>    i.e. :n:`@term’ = @term`.

Author: OK, `replace`.  The updates are not so simple to apply, especially after I've been at it for a while.

Shouldn't it be like this, not quite what you said?

![image](https://user-images.githubusercontent.com/1253341/101856815-eb6e6080-3b1a-11eb-8b45-0eea0f53e550.png)

>    i.e. :n:`@term’ = @term`.

Reviewer: LGTM.

   :n:`@occurrences`
     The `type of` and `value of` forms are not supported.

   :n:`by @ltac_expr3`
     Applies the :n:`@ltac_expr3` to solve the generated equality

   .. exn:: Terms do not have convertible types.
      :undocumented:

>    .. tacn:: cutrewrite {? {| -> | <- } } @one_term {? in @ident }

Author: It's immediately below the description of `replace`, whose second form mentions equalities.  You really think it's necessary to explain?

Also, in the description of `replace`, I think `assumption` should be `hypothesis`.

>    .. tacn:: cutrewrite {? {| -> | <- } } @one_term {? in @ident }

Reviewer: It doesn't behave at all like the second form of `replace`, where you provide a term that corresponds to the left-hand side (or right-hand side) of an equality. In the case of cutrewrite, you provide a term that corresponds to the full equality: `cutrewrite (X = Y)` can be replaced by `replace X with Y` (although as explained in #13480, the side-condition will not be in the same order, which is a problem which justifies not removing `cutrewrite` yet).

>    .. tacn:: cutrewrite {? {| -> | <- } } @one_term {? in @ident }

Author: > Also, in the description of `replace`, I think `assumption` should be `hypothesis`.

What about this?

>    .. tacn:: cutrewrite {? {| -> | <- } } @one_term {? in @ident }

Reviewer: > > Also, in the description of `replace`, I think `assumption` should be `hypothesis`.
> 
> What about this?

Yes, this was used as a synonym, but let's be consistent with the terminology and always use `hypothesis`.

     The :n:`@oriented_rewriter`\s are applied sequentially
     to the first goal generated by the previous :n:`@oriented_rewriter`.  If any of them fail,
     the tactic fails.

   :n:`{? {| -> | <- } }`
     For `->` (the default), :n:`term__1` is rewritten
     into :n:`term__2`.  For `<-`, :n:`term__2` is rewritten into :n:`term_1`.

   :n:`@natural`
     If followed by :n:`?`, gives the maximum number of rewrites that will be performed.
     Otherwise, it specifies the exact number of rewrites to perform (whether followed by
     `!` or not).  If the given number of rewrites can't be done, the tactic fails.

   :n:`?`
     Performs the rewrite as many times as possible (possibly zero times). This form never fails.

   :n:`!`
     Performs the rewrite as many times as possible and at least once.  Otherwise this
     form fails.

   :n:`@occurrences`
     If :n:`@occurrences` specifies multiple occurrences, the tactic succeeds if
     any of them can be rewritten.  If not specified, only the first occurrence
     in the conclusion is replaced.

     If :n:`at @occs_nums` is specified, rewriting is always done with setoid
     rewriting, even for Leibniz’s equality, so you must

>      ``Import Setoid`` to use this form.

Reviewer: Actually one only need to `Require Setoid`.

>      .. todo PR above is correct?  rewrite ... at 1 2 3 works even if only 1 is rewritten?

Reviewer: Apparently, yes:

```coq
Require Setoid.
Goal 1 = 0 -> 0 = 1.
intro H.
rewrite H at 1 2 3.
```

But shouldn't it be considered as a bug to fix (that it does not fail even if the occurrence doesn't exist)?

Author: Yup.  https://github.com/coq/coq/issues/13566

>    :n:`@ltac_expr3`

Reviewer: ```suggestion
   :n:`by @ltac_expr3`
```

     If specified, is used to resolve all side conditions generated by the tactic.

   .. exn:: The @term provided does not end with an equation.
      :undocumented:

   .. exn:: Tactic generated a subgoal identical to the original goal.
      :undocumented:

   .. exn:: Found no subterm matching @term in @ident.
            Found no subterm matching @term the current goal.

>       This happens if @term does not occur in, respectively, the named hypothesis or the goal.

Reviewer: ```suggestion
      This happens if :n:`@term` does not occur in, respectively, the named hypothesis or the goal.
```

   .. tacn:: erewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }

>       Works like :n:`rewrite @one_term`, but turns

Reviewer: ```suggestion
      Works like :tacn:`rewrite`, but turns
```

> file ``Logic.v`` (see :ref:`coq-library-logic`). The notation for :g:`eq T t u` is

Reviewer: This is inconsistent with the actual implicit arguments (`A` is implicit in the type of `eq`, not just in the notation). The paragraph could be slightly amended to be more consistent with the reality.

Author: Could you suggest wording?  I'm not sure what's needed.

simply :g:`t=u` dropping the implicit type of :g:`t` and :g:`u`.

.. tacn:: rewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }

   .. insertprodn oriented_rewriter constr_with_bindings_arg

   .. prodn::
      oriented_rewriter ::= {? {| -> | <- } } {? @natural } {? {| ? | ! } } @constr_with_bindings_arg

>       constr_with_bindings_arg ::= {? > } @one_term {? with @bindings }

Reviewer: I suggest calling this `term_with_bindings`. I would also like to understand what the `>` does in this case.

Author: How about `one_term_with_bindings`?

Reviewer: Sure.

   Rewrites terms based on the equality given by :n:`@one_term`, which
   must have the form:

>    :n:`forall (x__1: A__1) ... (x__n: A__n), eq term__1 term__2`

Reviewer: ```suggestion
   :n:`forall (x__1: A__1) ... (x__n: A__n), eq @term__1 @term__2`
```

>    where :g:`eq` is the Leibniz equality or a registered setoid equality.

Reviewer: We should say that, in the latter case, a different implementation is used, and link to the "Generalized rewriting" chapter for more about setoid rewriting.

Author: comme ça?

```
   where :g:`eq` is the Leibniz equality.  Use :tacn:`setoid_rewrite`, described in
   :ref:`generalizedrewriting`, when `eq` is a setoid equality.
```

>    :n:`rewrite @one_term` finds subterms matching `term`\ :sub:`1` in the goal,
>    and replaces them with `term`\ :sub:`2` using the values bound by the match

Reviewer: ```suggestion
   :n:`rewrite @one_term` finds subterms matching :n:`@term__1` in the goal,
   and replaces them with :n:`@term__2` using the values bound by the match
```

>    (or the reverse if `<-` is given).

Reviewer: I don't understand what "values bound by the match" mean. But in any case, the parenthesis should probably appear before.

Author: Matching uses unification, which binds values.  I was thinking about pattern variables, maybe that was the wrong idea.  I'll drop the clause.

   Some of the variables :g:`x`\ :sub:`i` are solved by unification,
   and some of the types :n:`A__1, ..., A__n` may become new
   subgoals.

   Rewriting is
   always done with setoid rewriting, even for Leibniz’s equality, so you must

>    ``Import Setoid`` to use this tactic.

Reviewer: This is only true of the variant using `at` AFAIU.

Author: I moved the statement under the `occurrences` description:

     If :n:`at @occ_nums` is specified, rewriting is always done with setoid
     rewriting, even for Leibniz’s equality, so you must
     ``Import Setoid`` to use this form.

   :n:`{+, @oriented_rewriter }`
     The :n:`@oriented_rewriter`\s are applied sequentially
     to the first goal generated by the previous :n:`@oriented_rewriter`.  If any of them fail,
     the tactic fails.

   :n:`@occurrences`
     If :n:`@occurrences` specifies multiple occurrences, the tactic succeeds if
     any of them can be rewritten.  If not specified, only the first occurrence

>      in the conclusion is replaced.

Reviewer: Should be moved below. We want to have finished describing the full syntax of `oriented_rewriter` first.

     .. todo PR above is correct?  rewrite ... at 1 2 3 works even if only 1 is rewritten?

   :n:`@ltac_expr3`

>      If specified, is used to resolve all side conditions generated by the tactic.

Reviewer: Should be moved below. We want to have finished describing the full syntax of `oriented_rewriter` first.

Author: I guess.  bfs order may work better for complex commands.

   :n:`{? {| -> | <- } }`
     For `->` (the default), :n:`term__1` is rewritten
     into :n:`term__2`.  For `<-`, :n:`term__2` is rewritten into :n:`term_1`.

   :n:`@natural`
     If followed by :n:`?`, gives the maximum number of rewrites that will be performed.
     Otherwise, it specifies the exact number of rewrites to perform (whether followed by
     `!` or not).  If the given number of rewrites can't be done, the tactic fails.

   :n:`?`
     Performs the rewrite as many times as possible (possibly zero times). This form never fails.

   :n:`!`
     Performs the rewrite as many times as possible and at least once.  Otherwise this
     form fails.

   .. todo PR: The following seems to be standard behavior for @occurrences, so I plan
      to drop (most of?) it.
      And I don't get the part about

>       "for all hypotheses :g:`H'` different from :g:`H`".  What do these mean?

Reviewer: You could also consider this to be an example, that could be added as such (`.. example::`) to the description of `occurrences`. I believe that the important piece of info here is the part that you don't understand: if you rewrite with one of your hypothesis in all of your hypotheses, the tactic won't attempt to rewrite in the hypothesis you are rewriting with.

Author: I think we need to mention that `one_term` can be a lemma or the name of a hypothesis, which the current text doesn't make clear.  And maybe in several other tactics.

Is `rewrite (contrapositive P Q) by dec` (from `FSetDecide.v`) is a setoid rewrite?  I don't `eq` in Logic.Decidable.contrapositive.  So not sure what to say just below the `tacn`..

I added some words under `occurrences`

Author: Please respond to the previous comment.

Reviewer: > I think we need to mention that `one_term` can be a lemma or the name of a hypothesis, which the current text doesn't make clear. And maybe in several other tactics.

I agree.

> Is `rewrite (contrapositive P Q) by dec` (from `FSetDecide.v`) is a setoid rewrite? I don't `eq` in Logic.Decidable.contrapositive. So not sure what to say just below the `tacn`.

Yeah, I think so. I'm not sure what determines whether setoid rewrite is used or not...

   + :n:`rewrite T in * |-` will do :n:`rewrite T in H` for all hypotheses
     :g:`H` different from :g:`T`.
     A success will happen as soon as at least one of these simpler tactics succeeds.

   .. exn:: The @term provided does not end with an equation.
      :undocumented:

>    .. exn:: Tactic generated a subgoal identical to the original goal. This happens if @term does not occur in the goal.

Reviewer: It looked like the second sentence was actually the description of the error message, not a part of it. But it is also an incorrect description AFAICT. The error message can occur in this case:

```coq
Parameter ax : 0 = 0.

Goal 0 = 1.
rewrite ax.
```

But when `term` does not occur in the goal, the usual error message is rather: "Found no subterm matching "..." in the current goal."

      the same key as the left- or right-hand side of the lemma given to rewrite, and the arguments
      are then unified up to full reduction.

.. tacn:: replace @one_term__from with @one_term__to {? @occurrences } {? by @ltac_expr3 }

   Replaces all free occurrences of :n:`@one_term__from`

>    in the current goal with :n:`@one_term__to’` and generates an equality

Reviewer: Left-over prime symbol:
```suggestion
   in the current goal with :n:`@one_term__to` and generates an equality
```

>    :n:`@one_term__from = @one_term__to’`

Reviewer: It was reported in #13480, that this documentation is actually incorrect. The generated equality is reversed.

   as a subgoal. This equality is automatically solved if it occurs among

>    the assumptions, or if its symmetric form occurs. It is equivalent to

Reviewer: I don't think this "It is equivalent to..." is really useful.

Author: I am so relieved :-)

   :n:`cut @one_term__from = @one_term__to’; [intro H`:sub:`n`:n:`;`
   :n:`rewrite <- H`:sub:`n` :n:`; clear H`:sub:`n`:n:`|| assumption || symmetry; try assumption]`.

   :n:`by @ltac_expr3`
     Applies the :n:`@ltac_expr3` to solve the generated equality

   :n:`@occurrences`

>      This may not contain `type of` or `value of``

Reviewer: I would rather say "The `type of` and `value of` forms are not supported."

   .. exn:: Terms do not have convertible types.
      :undocumented:

.. tacn:: subst {* @ident }

   For each :n:`@ident`, in order, for which there is a hypothesis in the form
   :n:`@ident = @term` or :n:`@term = @ident`, replaces :n:`@ident` with :n:`@term`
   everywhere in the hypotheses and the conclusion and clears :n:`@ident` and the hypothesis
   from the context.  If there are multiple hypotheses that match the :n:`@ident`,
   the first one is used.  If no :n:`@ident` is given, replacement is done for all
   hypotheses in the appropriate form in top to bottom order.

   If :n:`@ident` is a local definition of the form :n:`@ident := @term`, it is also
   unfolded and cleared.

   If :n:`@ident` is a section variable, it is expected to have no
   indirect occurrences in the goal, i.e. that no global declarations
   implicitly depending on the section variable are present in the

>    goal.

Reviewer: Could be moved in the description of the corresponding error message below.

   .. note::

>       + If the hypothesis is itself dependent in the goal, it is replaced by the proof of

Reviewer: Since there is only one bullet-point left over, the bullet could be removed.

        reflexivity of equality.

   .. flag:: Regular Subst Tactic

>       This flag controls the behavior of :tacn:`subst`. When it is

Reviewer: There are quite a few `:sub:` worth replacing with a regular `:n:` notation in the paragraph below.

          rewrite * {? {| -> | <- } } @one_term at @rewrite_occs in @ident {? by @ltac_expr3 }
   :name: rewrite *; _
   :undocumented:

.. tacn:: rewrite_db @ident {? in @ident }
   :undocumented:

.. tacn:: replace @one_term__from with @one_term__to {? @occurrences } {? by @ltac_expr3 }
          replace {? {| -> | <- } } @one_term__from {? @occurrences }
   :name: replace; _

   The first form replaces all free occurrences of :n:`@one_term__from`
   in the current goal with :n:`@one_term__to` and generates an equality
   :n:`@one_term__to = @one_term__from`
   as a subgoal. (Note the generated equality is reversed with respect
   to the order of the two terms in the tactic syntax; see
   issue `#13480 <https://github.com/coq/coq/issues/13480>`_.)
   This equality is automatically solved if it occurs among
   the hypotheses, or if its symmetric form occurs.

   The second form, with `->` or no arrow, replaces :n:`@one_term__from`
   with :n:`@term__to` using
   the first hypothesis whose type has the form :n:`@one_term__from = @term__to`.
   If `<-` is given, the tactic uses the first hypothesis with the reverse form,
   i.e. :n:`@term__to = @one_term__from`.

>    You must `Require Setoid` to use the tactic with a setoid equality.

Reviewer: As I said in my answer to your comment above, I don't think that this tactic can work with a setoid equality so this sentence should be removed. The one I wanted to add was that you must `Require Setoid` to use the `at` clause in `@occurrences`.

   :name: rewrite *; _
   :undocumented:

.. tacn:: rewrite_db @ident {? in @ident }
   :undocumented:

.. tacn:: replace @one_term__from with @one_term__to {? @occurrences } {? by @ltac_expr3 }
          replace {? {| -> | <- } } @one_term__from {? @occurrences }
   :name: replace; _

   The first form replaces all free occurrences of :n:`@one_term__from`
   in the current goal with :n:`@one_term__to` and generates an equality
   :n:`@one_term__to = @one_term__from`
   as a subgoal. (Note the generated equality is reversed with respect
   to the order of the two terms in the tactic syntax; see
   issue `#13480 <https://github.com/coq/coq/issues/13480>`_.)
   This equality is automatically solved if it occurs among
   the hypotheses, or if its symmetric form occurs.

   The second form, with `->` or no arrow, replaces :n:`@one_term__from`
   with :n:`@term__to` using
   the first hypothesis whose type has the form :n:`@one_term__from = @term__to`.
   If `<-` is given, the tactic uses the first hypothesis with the reverse form,
   i.e. :n:`@term__to = @one_term__from`.

   :n:`@occurrences`

>      The `type of` and `value of` forms are not supported.

Reviewer: I just discovered (by testing) that in the case of `replace` too, you need to `Require Setoid` to be able to use an `at` clause.

Author: I added a sentence saying that, but the tactic description doesn't otherwise mention setoid equality.  So I guess the part "and generates an equality :n:`@one_term__to = @one_term__from` as a subgoal" needs to be rephrased similar to the wording for `rewrite`, i.e. with `EQ`?

Reviewer: No, `replace` only works with Leibniz equality. The need to `Require Setoid` to be able to use an `at` clause seems to be a stupid limitation due to an implementation detail.

Reviewer 2: What's the implementation detail limitation?

Reviewer: The implementation detail is that `replace` uses `rewrite` internally and thus suffers from the same limitations (that the `at` clause is only implemented for `setoid_rewrite`).

Reviewer: ```coq
Require Setoid.
Set Typeclasses Debug.
Goal 0 = 0.
replace 0 with 1. (* no output *)
Undo.
replace 0 with 1 at 1. (* typeclass resolution output *)
```

      relation. Lemmas are expected to be in the form
      :g:`forall x y z, R x y -> eq y z -> R x z`.

   .. cmd:: Declare Right Step @one_term

       Adds :n:`@term` to the database used by :tacn:`stepr`.

Rewriting with definitional equality
------------------------------------

.. tacn:: change {? @one_term__from {? at @occs_nums } with } @one_term__to {? @occurrences }

   Replaces terms with other :term:`convertible` terms.
   If :n:`@one_term__from` is not specified, then :n:`@one_term__from` replaces the conclusion and/or
   the specified hypotheses.  If :n:`@one_term__from` is specified, the tactic replaces occurrences
   of :n:`@one_term__to` within the conclusion and/or the specified hypotheses.

   :n:`{? @one_term__from {? at @occs_nums } with }`
     Replaces the occurrences of :n:`@one_term__from` specified by :n:`@occs_nums`
     with :n:`@one_term__to`, provided that the two :n:`@one_term`\s are
     convertible.  :n:`@one_term__from` may contain pattern variables such as `?x`,
     whose value which will substituted for `x` in :n:`@one_term__to`, such as in
     `change (f ?x ?y) with (g (x, y))` or `change (fun x => ?f x) with f`.

>    :n:`@occurrences`
>      If `with` is not specified, :n:`@occurrences` must only specify
>      entire hypotheses and/or the goal; it must not include any
>      :n:`at @occs_nums` clauses.

Reviewer: Should we add the corresponding error message here or below with the other error messages?

```suggestion
   :n:`@occurrences`
     If `with` is not specified, :n:`@occurrences` must only specify
     entire hypotheses and/or the goal; it must not include any
     :n:`at @occs_nums` clauses.
     
     .. exn:: Found an "at" clause without "with" clause.
        :undocumented:
```

Author: I put it after the other two error messages.  Indenting it under "occurrences" might look a bit like a formatting error.  I'm not convinced that the current approach to showing error messages is all that great--a lot of messages apply to many commands/tactics.  A topic for another day.

   .. exn:: Not convertible.
      :undocumented:

>    .. exn:: Too few occurrences.
>       :undocumented:

Reviewer: After grepping for this error message, I conclude that it does not exist anymore.
```suggestion
```

   .. tacn:: now_show @one_term

>       This is a synonym for :n:`change @conversion`. It can be used to

Reviewer: ```suggestion
      This is a synonym for :n:`change @one_term`. It can be used to
```

BTW, you should remove `conversion` from the `nitpick_ignore` list. It would have helped to catch this issue. And it's also time to remove `oriented_rewriter` from this list as well.

      make some proof steps explicit when refactoring a proof script
      to make it readable.

   .. seealso:: :ref:`Performing computations <performingcomputations>`

.. tacn:: change_no_check {? @one_term__from {? at @occs_nums } with } @one_term__to {? @occurrences }

>    For advanced usage. Similar to :tacn:`change` :n:`@term`, but as an optimization,

Reviewer: ```suggestion
   For advanced usage. Similar to :tacn:`change`, but as an optimization,
```

   it skips checking that :n:`@one_term__to` is convertible with the goal or
   :n:`@one_term__from`.

   Recall that the Coq kernel typechecks proofs again when they are concluded to
   ensure correctness. Hence, using :tacn:`change` checks convertibility twice
   overall, while :tacn:`change_no_check` can produce ill-typed terms,
   but checks convertibility only once.
   Hence, :tacn:`change_no_check` can be useful to speed up certain proof
   scripts, especially if one knows by construction that the argument is
   indeed convertible to the goal.

   In the following example, :tacn:`change_no_check` replaces :g:`False` with
   :g:`True`, but :cmd:`Qed` then rejects the proof, ensuring consistency.

   .. example::

>       .. coqtop:: all abort
> 
>          Goal False.
>            change_no_check True.
>            exact I.
>          Fail Qed.
> 
>    .. example::
> 
>       .. coqtop:: all abort
> 
>          Goal True -> False.
>            intro H.
>            change_no_check False in H.
>            exact H.
>          Fail Qed.

Reviewer: ```suggestion
      .. coqtop:: all abort fail

         Goal False.
           change_no_check True.
           exact I.
         Qed.

   .. example::

      .. coqtop:: all abort fail

         Goal True -> False.
           intro H.
           change_no_check False in H.
           exact H.
         Qed.
```

   .. tacn:: convert_concl_no_check @one_term

>       .. todo PR: Can we remove this now?

Reviewer: Yes, we should be able to remove it in 8.14. However, since this PR is targeting 8.13, we should keep the documentation until the removal. Thus, it would be better to wait for this PR to be merged before opening the PR to remove it.

   then replaces the goal by :n:`R @term @term` and adds a new goal stating
   :n:`eq @term @term`.

   If :n:`@ltac_expr` is specified, it is applied to the second goal after :tacn:`stepl`.

   .. cmd:: Declare Left Step @one_term

      Adds :n:`@one_term` to the database used by :tacn:`stepl`.

   This tactic is especially useful for parametric setoids which are not accepted
   as regular setoids for :tacn:`rewrite` and :tacn:`setoid_replace` (see
   :ref:`Generalizedrewriting`).

   .. tacn:: stepr @one_term {? by @ltac_expr }

      This behaves like :tacn:`stepl` but on the right hand side of the binary
      relation. Lemmas are expected to be in the form
      :g:`forall x y z, R x y -> eq y z -> R x z`.

   .. cmd:: Declare Right Step @one_term

       Adds :n:`@term` to the database used by :tacn:`stepr`.

Performance-oriented tactic variants
------------------------------------

.. tacn:: change @conversion {? @occurrences }

   .. insertprodn conversion conversion

   .. prodn::
      conversion ::= @one_term {? {? at @occs_nums } with @one_term }

   .. todo PR: need explanation of conversion in a couple tactics, should share

>       no idea what the "with" clause does

Author: I thought about splicing, but `conversion` is need to compare to `ltac2_conversion` in the Ltac2 chapter.

...

>       no idea what the "with" clause does

Reviewer: OK, that makes sense, I guess.

   Implements the rule ``Conv`` given in
   :ref:`subtyping-rules`. :g:`change U` replaces the current goal `T`
   with `U` provided that `U` is well-formed and that `T` and `U` are
   convertible.

   :n:`{? at @occs_nums } with @one_term__repl`
     Replaces the occurrences specified by :n:`@occs_nums` of :n:`@one_term`
     with :n:`@one_term__repl`.

   .. exn:: Not convertible.
      :undocumented:

   .. exn:: Too few occurrences.
      :undocumented:

   .. todo PR: Need to integrate the following elsewhere.  But how are
      the occs_nums used?  Should the syntax be improved, seems a little confusing.
      Maybe ask @herbelin?

   .. tacv:: change @one_term {? {? at {+ @occs_nums } } with @one_term__repl } {? @occurrences }

      In the presence of :n:`with`, this applies :tacn:`change` to the
      occurrences specified by :n:`@occurrences`. In the
      absence of :n:`with`, :n:`@occurrences` is expected to
      only list hypotheses (and optionally the conclusion) without
      specifying occurrences (i.e. no :n:`at` clause).

>    .. tacn:: now_show @conversion

Reviewer: Actually `now_show` can only take a `one_term` as argument.

Author: Why do you say that?  I see this:

```
| "change" conversion clause_dft_concl

Ltac now_show c := change c.
```

Reviewer: I think your misunderstanding what `Ltac` does. It doesn't pass everything it gets syntactically, this `Ltac` takes a single argument and thus doesn't use the `with` clause of `change`.

Author: Ah, OK.  Easy to get confused since Ltac semantics are vague, also too many ways to define tactics, all with different capabilities and quirks.  And constantly switching between them all.

>       dependencies in the goal.

Reviewer: In the error messages above, the `:n:` should be removed since the notation syntax is already activated by default (as in commands).

      hypotheses, which without the flag it may break.

>       default.

Author: I removed `default.`.  Did you mean more than that?

   database of lemmas of the form :g:`forall x y z, R x y -> eq x z -> R z y`
   where `eq` is typically a setoid equality. The application of :n:`stepl @one_term`
   then replaces the goal by :n:`R @one_term @term__2` and adds a new goal stating
   :n:`eq @one_term @term__1`.

   If :n:`@ltac_expr` is specified, it is applied to the side condition.

   .. cmd:: Declare Left Step @one_term

      Adds :n:`@one_term` to the database used by :tacn:`stepl`.

   This tactic is especially useful for parametric setoids which are not accepted
   as regular setoids for :tacn:`rewrite` and :tacn:`setoid_replace` (see
   :ref:`Generalizedrewriting`).

   .. tacn:: stepr @one_term {? by @ltac_expr }

      This behaves like :tacn:`stepl` but on the right hand side of the binary
      relation. Lemmas are expected to be in the form
      :g:`forall x y z, R x y -> eq y z -> R x z`.

   .. cmd:: Declare Right Step @one_term

       Adds :n:`@term` to the database used by :tacn:`stepr`.

Rewriting with definitional equality
------------------------------------

.. tacn:: change @conversion {? @occurrences }

   .. insertprodn conversion conversion

   .. prodn::

>       conversion ::= @one_term {? {? at @occs_nums } with @one_term }

Reviewer: OK then, let's not splice `conversion`. But then, I propose to transform it so that it appears as two separate cases, like this:
```suggestion
      conversion ::= @one_term__to
      | @one_term__from {? at @occs_nums } with @one_term__to
```

Another option, with a single case is this:
```suggestion
      conversion ::= {? @one_term__from {? at @occs_nums } with } @one_term__to
```

The goal of this change being to highlight that (contrary to `replace`), the `one_term__to` must always be provided and it is the `one_term__from` that may be omitted (in which case the whole goal is replaced).

Author: The only way to do this is to remove the `insertprodn`, which means it won't be updated by `doc_grammar`.  I'll splice it.

Reviewer: What!? No!

I was just suggesting an edit to add to `common.edit_mlg`, like we used to do all the time.

Author: Can you be more specific?  No to splicing it?  That was your original suggestion.

What edit do you want?  Specifying subscripts from `common.edit_mlg` is not a good idea, `doc_grammar` is really not set up to support that.

Reviewer: OK my mistake. I misunderstood your comment and didn't realize that you cannot specify subscripts in a `prodn` (would it be hard to add?). It looks good like this.

Author: At this point I'm reluctant to add anything to `doc_grammar` unless it's really necessary or quite simple.  It might take a couple days to do this.

For various reasons, the syntax updates are taking a very long time (long delays), which is discouraging.  It indicates that any significant improvements to Coq are likely to take a very long time to complete and really undercuts the motivation to undertake new improvements.

Author: There ought to be an aphorism something like "Gratification deferred is gratification denied".

Author: Or let me put it this way: I don't want to make such a change a prerequisite to completing this PR.

Reviewer: It was just a suggestion, not a request, and certainly not a prerequisite for this PR.

However, since you complain about your doc PRs taking a very long time to merge, I'll respond. I've spent considerable more time reviewing PRs (yours) this year that I had ever before. I do this because I see the value but I don't have unlimited time to allocate to this task, so it takes time and there are delays. You're opening PRs containing very large changes. This is not the good strategy if you want rapid reviews and merging. Your syntax updates PRs are usually not in a great shape when you open them (descriptions do not match anymore with the documented syntax) so it's normal that many feedback cycles are needed to get them ready. (And of course, the time zone difference between the two of us is also significantly delaying these feedback cycles even when I'm the most available, like during the first lockdown.)

Author: My comment was not meant as a complaint directed at you.  I'm sorry if my wording gave that impression.  Like everyone else, you're busy with many other things and your participation on any given effort is voluntary.  Your detailed reviews have been indispensable and very much appreciated.

I agree it's a challenging process getting these PRs right.  I'm open to suggestions to make it go more smoothly--though I think much of the challenge is probably inherent rather than due to our use of an inadequate process.

The current form of `doc_grammar` is set up to update an entire rst file, which makes it a natural increment for PRs.  That's too much to do at once for the tactics chapter (3000 lines), SSR (5800 lines) and maybe too much for the rewriting chapter (800 lines).  We could take an approach in which we do full updates on, say, the first 400 lines.  When that part is ready to commit, I can create a second branch that keeps all the changes.  Then, on the first branch, I remove all the changes made in `rewriting.rst` outside those 400 lines before we commit the changes (with possible exceptions in some cases, such as for material moved from the end into the 400 lines).   
 Then do the rest of the file in a new PR.  Do you want to try that for this PR?

Reviewer: >much of the challenge is probably inherent

Agreed.

>Do you want to try that for this PR?

:+1: Since I never managed to push my review beyond the documentation of `change`, it would make sense to postpone the rest to a later PR. This would also help ensure that this PR gets merged in time for 8.13.0.

Author: OK, we'll make this one cover everything before `Performing computations`.
